File: android/src/main/kotlin/com/baseflow/permissionhandler/utils/Codec.kt
package com.baseflow.permissionhandler.utils

import com.baseflow.permissionhandler.data.PermissionGroup
import com.baseflow.permissionhandler.data.PermissionStatus
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken

class Codec {
    companion object {
        @JvmStatic
        private val gsonDecoder : Gson = GsonBuilder().enableComplexMapKeySerialization().create()

        @JvmStatic
        fun decodePermissionGroup(arguments: Any) : PermissionGroup {
            return Codec.gsonDecoder.fromJson(arguments.toString(), PermissionGroup::class.java)
        }

        @JvmStatic
        fun decodePermissionGroups(arguments: Any) : Array<PermissionGroup> {

            var permissionGroupsType = object: TypeToken<Array<PermissionGroup>>() {}.type
            return Codec.gsonDecoder.fromJson(arguments.toString(), permissionGroupsType)
        }

        @JvmStatic
        fun encodePermissionStatus(permissionStatus: PermissionStatus) : String {
            return gsonDecoder.toJson(permissionStatus)
        }

        @JvmStatic
        fun encodePermissionRequestResult(permissionResults: Map<PermissionGroup, PermissionStatus>) : String {
            val jsonString = gsonDecoder.toJson(permissionResults)
            return jsonString
        }
    }
}

File: ios/Classes/utils/Codec.swift
//
//  Codec.swift
//  permission_handler
//
//  Created by Maurits van Beusekom on 25/07/2018.
//

import Foundation

struct Codec {
    private static let jsonDecoder = JSONDecoder()
    private static let jsonEncoder = JSONEncoder()
    
    static func decodePermissionGroup(from arguments: Any?) -> PermissionGroup {
        var permissionString: String = arguments as! String
        permissionString.removeFirst()
        permissionString.removeLast()
        
        return PermissionGroup(rawValue: permissionString)!
    }
    
    static func decodePermissionGroups(from arguments: Any?) -> [PermissionGroup] {
        let data = (arguments as! String).data(using: .utf8)
        let permissions = try! jsonDecoder.decode([PermissionGroup].self, from: data!)

        return permissions
    }
    
    static func encodePermissionStatus(permissionStatus: PermissionStatus) -> String? {
        let status = "\"" + permissionStatus.rawValue + "\""
        return status
    }
    
    static func encodeServiceStatus(serviceStatus: ServiceStatus) -> String? {
        let status = "\"" + serviceStatus.rawValue + "\""
        return status
    }
    
    static func encodePermissionRequestResult(permissionStatusResult: [PermissionGroup: PermissionStatus]) -> String? {
        let jsonDict = Dictionary(uniqueKeysWithValues:
            permissionStatusResult.map {
                (key: PermissionGroup, value: PermissionStatus) in (key.rawValue, value.rawValue)
                
        })
        
        let jsonData = try! JSONSerialization.data(withJSONObject: jsonDict, options: [])
        let jsonString = String(data: jsonData, encoding: .utf8)!

        return jsonString
    }
}
